### ISSUE TITLE
docker service ls | Error response from daemon: rpc error: code = ResourceExhausted desc = grpc: received message larger than max (5097134 vs. 4194304)

### ISSUE THREAD

**Issue by fabiopedrosa** [moby/moby#37997]
**Description**

`docker service ls`

fails with 

`Error response from daemon: rpc error: code = ResourceExhausted desc = grpc: received message larger than max (5097134 vs. 4194304)`

after upgrading from docker 18.04 to 18.06.1-ce

**Output of `docker version`:**

```
Client:
 Version:           18.06.1-ce
 API version:       1.38
 Go version:        go1.10.3
 Git commit:        e68fc7a
 Built:             Tue Aug 21 17:24:56 2018
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.06.1-ce
  API version:      1.38 (minimum version 1.12)
  Go version:       go1.10.3
  Git commit:       e68fc7a
  Built:            Tue Aug 21 17:23:21 2018
  OS/Arch:          linux/amd64
  Experimental:     false
```

**Output of `docker info`:**

```
Containers: 102
 Running: 102
 Paused: 0
 Stopped: 0
Images: 2636
Server Version: 18.06.1-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 1422
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: bcmt7u1pea2jsixbgik4rsro3
 Is Manager: true
 ClusterID: f95tt2krxxhhmxli0mnj1umlx
 Managers: 9
 Nodes: 9
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Autolock Managers: false
 Root Rotation In Progress: false
 Node Address: 192.168.51.107
 Manager Addresses:
  192.168.51.107:2377
  192.168.51.108:2377
  192.168.51.109:2377
  192.168.51.110:2377
  192.168.51.121:2377
  192.168.51.122:2377
  192.168.51.123:2377
  192.168.51.124:2377
  192.168.51.125:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 468a545b9edcd5932818eb9de8e72413e616e86e
runc version: 69663f0bd4b60df09991c08812a60108003fa340
init version: fec3683
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-127-generic
Operating System: Ubuntu 16.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 12
Total Memory: 62.81GiB
Name: pf-us1-sm2
ID: OZK4:WPBM:L64X:6WXK:QDB7:UBZF:YTKA:MC3X:GNDK:DWQM:KGJ4:M24E
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 192.168.51.107:5000
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Comment by olljanat** [comment]
@fabiopedrosa that is interesting. How many service approximately you have running on there?

Are you able to use ```docker service logs <service name>``` or ```docker service inspect <service name>``` commands?

Can you try either restart leading manager or demote it to worker (if you have important workloads running on it). You can find it using command: ```docker node ls```

Finally if these does not help then I would suggest restart manager nodes one by on e (and this point it is nice if you have dedicated manager nodes which does not have important workloads).

**Comment by fabiopedrosa** [comment]
@olljanat We have 20 services, and around 500-600 containers running across 9 nodes. Average of 60 per node, which should be a low number. We've run a lot more in previous versions of docker.

Both `docker service inspect` and `docker service logs <name>` work fine.

I've tried restarting the docker service in the current leader, and it automatically made another node the leader. But the issue remained the same.

I'll try restarting each manager one by one at a better time (off-peak).

**Comment by dperny** [comment]
it's not hitting the limit because of the number of services. the CLI operation for `docker service ls` includes a call to the task list endpoint, which in turns calls swarmkit to get all tasks, and there are too many tasks to include in one response.

we need to paginate task list responses such that they don't get chomped by the GRPC message size limit. or, otherwise, we need to remove the limitation on GRPC message size for swarmkit API responses (i believe the message size limitation exists for reasons related to passing raft messages between members)

(probably, i might be wrong. this is a 60 second educated guess at the underlying problem.)

**Comment by olljanat** [comment]
@fabiopedrosa @xdexter one thing which comes to my mind is that do you have some service which is constantly failing and gets restarted?

If so it might help if you include **--restart-max-attempts 5** setting for those services. Then new tasks are not created all the time and swarm is able to do clean up go old tasks.

**Comment by fabiopedrosa** [comment]
@olljanat No, that's not the case for us. We had no containers stop/fail recently, and even made sure to run `docker prune container -a -y` on all nodes before creating this ticket.

**Comment by xdexter** [comment]
@olljanat For us not too, all services has this config:

> restart_policy:
  delay: 60s
  max_attempts: 10

The error started when i updated my docker to "18.06.1-ce", in older version (17.03.1-ce) the problem did not happen.

**Comment by olljanat** [comment]
Ok. You probably can use ```--filter``` as workaround or downgrade to 18.03.1: https://docs.docker.com/engine/reference/commandline/service_ls/#filtering

**Comment by fabiopedrosa** [comment]
@olljanat I understand how that can help, but I hope you guys consider either implementing pagination, allow for a custom message max_size or just increase that limit by default.

**Comment by olljanat** [comment]
@fabiopedrosa I'm not Docker maintainer so I cannot say anything about when or how this will be fixed. But I just try ask rights questions to get as much as possible useful information to here and tell about possible workarounds as much I understand this stuff.

**Comment by xdexter** [comment]
@olljanat hmm no, i already tried add filter to command but the same error happen. @fabiopedrosa i agree with you!

**Comment by jmcdonagh** [comment]
Having the same issue with about 200 services, 3 containers a piece approximately.

**Comment by rveede** [comment]
Same here. Just over 200 services with most of them having 3 replicas.

**Comment by kjelle** [comment]
I just https://github.com/moby/moby/issues/37941#issuecomment-433307297 - we have messages on ~319 MB...

**Comment by cpuguy83** [comment]
Does lowering `--snapshot-interval` on `docker swarm update` help? Potentially also `--task-history-limit`.

**Comment by cpuguy83** [comment]
Looks like this regression is possibly caused by a change in grpc to set the message size limit to 4MB from a very high limit (math.MaxInt32). See the referenced PR above.

Although huge message sizes seems like another thing that needs to be resolved.
